### PR TITLE
Port git-dit-new

### LINKED ISSUE #37: Port "git-dit-new" to rust
What the title says. Porcelain commands should go in the binary.

### PR DESCRIPTION
**Pull request by neithernut**
Resolves #37.

### CODE REVIEW

**Comment by neithernut** [comment]
The thing is not tested, yet.

File: src/main.rs [hunk 1]
```diff
@@ -143,7 +144,42 @@ fn get_issue_tree_init_hashes(repo: &Repository, _: &clap::ArgMatches) -> i32 {
...
+/// new subcommand implementation
+///
+fn new_impl(repo: &Repository, matches: &clap::ArgMatches) -> i32 {
+    let sig = try_or_1!(repo.signature());
+
+    // get the message, either from the command line argument or an editor
+    let message = if let Some(paragraphs) = matches.values_of("message") {
+        // the message was supplied via the command line
+        paragraphs.map(str::to_owned).map(|p| (p + "\n").to_owned()) // paragraphs
+                  .chain(try_or_1!(repo.prepare_trailers(matches)).into_iter().map(|trailer| trailer.to_string())) // TODO: figure out how to write `ToString::to_string` as a closure here (dragged in through `Display`)
```
Comment (neithernut) [inline src/main.rs]: @matthiasbeyer before you ask: `ToString::to_string()` takes a `&self`. Hence, you cannot simply write `.map(Trailer::to_string)`.

File: src/main.rs [hunk 2]
```diff
@@ -143,7 +144,42 @@ fn get_issue_tree_init_hashes(repo: &Repository, _: &clap::ArgMatches) -> i32 {
...
+/// new subcommand implementation
+///
+fn new_impl(repo: &Repository, matches: &clap::ArgMatches) -> i32 {
+    let sig = try_or_1!(repo.signature());
+
+    // get the message, either from the command line argument or an editor
+    let message = if let Some(paragraphs) = matches.values_of("message") {
+        // the message was supplied via the command line
+        paragraphs.map(str::to_owned).map(|p| (p + "\n").to_owned()) // paragraphs
+                  .chain(try_or_1!(repo.prepare_trailers(matches)).into_iter().map(|t| t.to_string()))
```
Comment (neithernut) [inline src/main.rs]: @matthiasbeyer before you ask: you cannot simply write `.map(Trailer::to_string)` because `ToString::to_string()` takes a `&self`.

**Comment by neithernut** [comment]
Tested superficially.

File: src/cli.yaml [hunk 3]
```diff
@@ -161,8 +161,10 @@ subcommands:
             - message:
                 short: m
                 long: message
-                help: Use this as issue message
-                multiple: false
+                help: >
+                        Use this as issue message. If used multiple times, each
+                        argument will be a paragraph in the comit message.
```
Comment (matthiasbeyer) [inline src/cli.yaml]: `s/comit/commit/`

File: src/error.rs [hunk 4]
```diff
@@ -38,5 +38,10 @@ error_chain! {
             description("Could not find some configuration or ENV variable specifying a program")
             display("Could not find {} configuration or ENV variable", program_name)
         }
+
+        ChildError {
+            description("A child program was unsucessful")
+            display("A child program was unsucessful")
```
Comment (matthiasbeyer) [inline src/error.rs]: `s/unsucessful/unsuccessful/`

File: src/error.rs [hunk 5]
```diff
@@ -38,5 +38,10 @@ error_chain! {
             description("Could not find some configuration or ENV variable specifying a program")
             display("Could not find {} configuration or ENV variable", program_name)
         }
+
+        ChildError {
+            description("A child program was unsucessful")
```
Comment (matthiasbeyer) [inline src/error.rs]: `s/unsucessful/unsuccessful/`

File: src/main.rs [hunk 6]
```diff
@@ -143,7 +144,42 @@ fn get_issue_tree_init_hashes(repo: &Repository, _: &clap::ArgMatches) -> i32 {
...
-// ...
+/// new subcommand implementation
+///
+fn new_impl(repo: &Repository, matches: &clap::ArgMatches) -> i32 {
+    let sig = try_or_1!(repo.signature());
+
+    // get the message, either from the command line argument or an editor
+    let message = if let Some(paragraphs) = matches.values_of("message") {
+        // the message was supplied via the command line
+        paragraphs.map(str::to_owned).map(|p| (p + "\n").to_owned()) // paragraphs
```
Comment (matthiasbeyer) [inline src/main.rs]: Would like to see a newline before the second `map`.

File: src/util.rs [hunk 7]
```diff
@@ -54,5 +82,52 @@ impl<'r> RepositoryUtil<'r> for Repository {
...
     }
+
+    fn commitmsg_edit_path(&self) -> PathBuf {
+        self.path().with_file_name("COMMIT_EDITMSG")
+    }
+
+    fn get_commit_msg(&self, path: PathBuf) -> Result<Vec<String>> {
+        // let the user write the message
+        if !run_editor(self.config().chain_err(|| EK::WrappedGitError)?, &path)?
+            .wait().chain_err(|| EK::WrappedIOError)?
```
Comment (matthiasbeyer) [inline src/util.rs]: Newline after `wait()` call?

Comment (neithernut) [inline src/util.rs]: The `chain_err()` belongs to the `wait()`. Consider a chain such as:
```
.foo().chain_err(...)?
.bar().chain_err(...)?
```

Writing it as
```
.foo()
.chain_err(...)?
.bar()
.chain_err(...)?
```
may confuse readers.

**Pull request merged**